lib/gitsh/tab_completion/automaton.rb

```diff
@@ -35,19 +35,32 @@ class State
         def initialize(name)
           @name = name
           @transitions = Hash.new { |h, k| h[k] = Set.new }
+          @fallback_transitions = Hash.new { |h, k| h[k] = Set.new }
           @free_transitions = Set.new
         end
 
         def add_transition(matcher, state)
           transitions[matcher] << state
         end
 
+        def add_fallback_transition(matcher, state)
+          fallback_transitions[matcher] << state
+        end
+
         def add_free_transition(state)
           free_transitions << state
         end
 
         def follow(token)
-          matching_transitions(token).map(&:freely_reachable).inject(Set.new, :|)
+          new_states = matching_transitions(token).
+            map(&:freely_reachable).
+            inject(Set.new, :|)
+
+          if new_states.empty?
+            fallback_transitions.values.inject(Set.new, :|)
```

Comment:
There's a potential bug here: we're not calling `freely_reachable` on the resulting states, so a fallback transition followed by a free transition won't work.

Reply:
Fixed in ad6a4d0